[doc] Add glossary of terms

jwnrt opened the pull request:
This PR adds a glossary of terms to the documentation. Resolves https://github.com/lowRISC/opentitan/issues/19437.

Many of these entries only state the expansion of an acronym without defining what it means.

Please feel free to suggest changes to explanations or for new entries.

Future PRs:

- Add short explanations of what acronyms actually mean.
- Link to the glossary from terms used around the documentation.
- Consolidate glossaries found in block-level documentation into this page.

cdgori reviewed (commented):
This is a great start!
 
I'd probably add: CW340, Darjeeling, FIPS, GF2 (or GF(2)), SHA

Some of the block-level MDs have their own glossary sections as well, if we want to unify them(?)

jwnrt commented:
> I'd probably add: CW340, Darjeeling, FIPS, GF2 (or GF(2)), SHA

Thanks for the good suggestions, I've added these

I've also switched the glossary Markdown to just using titles and paragraphs because:

* They render much nicer by default.
* They come with IDs by default, so we can use just Markdown and no HTML.
* We get highlighting and the TOC that mdbook generates for section headers.

> Some of the block-level MDs have their own glossary sections as well, if we want to unify them(?)

Yeah, I think that will be a good idea. I'll move their descriptions in when we add links to the glossary across the docs.

HU90m commented:
> Yeah, I think that will be a good idea. I'll move their descriptions in when we add links to the glossary across the docs.

We could also think about categorizing into sections when we do this. I can see the glossary can getting quite long.

jwnrt commented:
> We could also think about categorizing into sections when we do this. I can see the glossary can getting quite long.

Sure, the TOC on the right definitely helps when it's this long

Any idea what categories you'd want? It might be difficult separate them cleanly. Maybe split into project-specific / general?

HU90m commented:
> Any idea what categories you'd want? It might be difficult separate them cleanly. Maybe split into project-specific / general?

I'm not sure. I think we can probably work this out at a later date.

hcallahan-lowrisc commented on doc/glossary.md:
```diff
@@ -0,0 +1,313 @@
+
+Design Verification.
+
+#### DVsim
```
We almost-exclusively use this acronym as lower-case "dvsim" throughout the codebase.
I think it's probably worth doing the same here?

hcallahan-lowrisc commented on doc/glossary.md:
```diff
@@ -0,0 +1,313 @@
+
+#### EDA
+
+Electronic Design Automation (tool).
```
I would probably drop '(tool)', as the acryonym is sometimes used to talk about a process more-generally than a specific tool

hcallahan-lowrisc commented on doc/glossary.md:
```diff
@@ -0,0 +1,313 @@
+
+Enhanced Physical Memory Protection.
+
+A RISC-V extension to the physical memory protection feature.
```
```suggestion
A RISC-V extension to the physical memory protection part of the RISC-V Privileged Architecture Specification.
```

hcallahan-lowrisc commented on doc/glossary.md:
```diff
@@ -0,0 +1,313 @@
+
+#### Flash
+
+Non-volatile reprogrammable memory.
```
```suggestion
A type of non-volatile reprogrammable memory.
```

hcallahan-lowrisc commented on doc/glossary.md:
```diff
@@ -0,0 +1,313 @@
+
+#### HyperDebug
+
+Custom development board for testing and debugging OpenTitan devices.
```
```suggestion
Custom OpenTitan development harness for testing and debugging deployments on both CW310 and CW340 boards.
```

hcallahan-lowrisc commented on doc/glossary.md:
```diff
@@ -0,0 +1,313 @@
+
+#### OTBN
+
+OpenTitan Big Number - a RISC-V-like big number accelerator core.
```
```suggestion
OpenTitan Big Number Accelerator - a RISC-V-like programmable coprocessor for asymmetric cryptographic algorithms.
```
From : https://opentitan.org/book/hw/doc/cores.html

hcallahan-lowrisc commented on doc/glossary.md:
```diff
@@ -0,0 +1,313 @@
+
+#### OTP
+
+One-Time Programmable memory.
```
```suggestion
One-Time Programmable (memory).
```

hcallahan-lowrisc commented on doc/glossary.md:
```diff
@@ -0,0 +1,313 @@
+#### RISC-V
+
+The open source instruction set architecture used for OpenTitan cores.
+
```
```suggestion
Suggestion: RNG
```

hcallahan-lowrisc commented on doc/glossary.md:
```diff
@@ -0,0 +1,313 @@
+
+#### Silicon
+
+The metal semiconductor material that most chips are made from.
```
```suggestion
The semiconductor material that most microchips are made from.
```

hcallahan-lowrisc commented on doc/glossary.md:
```diff
@@ -0,0 +1,313 @@
+
+Static Random-Access Memory.
+
+#### Stapping
```
```suggestion
#### Strap/Strapping
```

hcallahan-lowrisc commented on doc/glossary.md:
```diff
@@ -0,0 +1,313 @@
+
+#### Tapeout
+
+The process of fabricating a chip onto silicon.
```
```suggestion
The process of fabricating a chip.
```

hcallahan-lowrisc commented on doc/glossary.md:
```diff
@@ -0,0 +1,313 @@
+
+#### UART
+
+Universal asynchronous receiver-transmitter.
```
```suggestion
Universal Asynchronous Receiver-Transmitter.
```

jwnrt commented on doc/glossary.md:
```diff
@@ -0,0 +1,313 @@
+
+#### Stapping
+
+Configuration of GPIO input settings used to change the state of a chip.
```
Yeah, this was a hard one.

How about:

> GPIO signals sampled by OpenTitan for configuration.

jwnrt replied:
I'll use this, but FYI some lifecycle states sample straps continuously so they're not only for early boot.

PS @HU90m, our version of mdbook [has this feature](https://github.com/rust-lang/mdBook/blob/master/CHANGELOG.md#mdbook-0430), but are you happy for me to use it style-wise?

```markdown
#### Configuration Straps / Pin Straps {#straps}
```

This sets the ID of the header so that links can be written shorter. I'm not adding links in this PR, so will leave it out for now. I'm not sure how GitHub handles these.

hcallahan-lowrisc replied:
We should probably come up with a fuller definition in time, as straps don't seem to be well-defined elsewhere in the repo. In general, they exist to be a simple way to configure the system, that is typically static (pull up/down with a resistor) but less frequently is dynamic. They are one of the only ways to provide early-boot configuration (though only a small number of bits of config), but the pin/connection can then often be used/reused for other things e.g. GPIO.

jwnrt replied:
Sure, happy to defer this. I've taken your definition for now anyway.
